keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1190 @@
+
+##### Blacklisted Validations
+
+To mitigate this, validation failures in rules which have an
+ancestor of one of these types will not be ratcheted:
+
+- not: Supporting negation of rules for ratcheting would require a complete 
+rewrite of the validation system.
+- oneOf: Requires EXACTLY one alternative, so it is not allowed for the same reason as `not`  rules.
+- x-kubernetes-validations which make use of `oldSelf`: To get ratcheting to work in an intuitive way with transition rules would requiring the resource two states back. This is
+not feasible. Also, users can write ratcheting logic directly in CEL as a 
+transition rule if that was required.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+At a high level, CR validation in kubernetes has three components:
+1. `kube-openapi`-based `Validate` using `*spec.Schema`
+2. StructuralSchema-based Meta, TypeMeta, ListSet/Map validation
+3. CEL Validators using a CEL schema interface type
+
+Each of these validation paths will need to have their `ValidateUpdate(old, new)`
+codepath modified to:
+
+1. Take an option to enable the feature
+2. Ignore errors which occur both on the `old` and `new` objects
+3. Return ignored errors in a separate list
+
+### `kube-openapi` changes
+
+The `kube-openapi` schema validator validates the standard OpenAPI/JSONSchema
+elements of the CRD schema. This is basically everything except CEL rules,
+and listset/map types.
+
+The current SchemaValidator will need to be updated to add a new sibling to
+its [`Validate(obj) *Result`](https://github.com/kubernetes/kube-openapi/blob/7562a10069617a0c05535f22bf1561b1bce67869/pkg/validation/validate/schema.go#LL109C46-L109C46) method: `ValidateUpdate(old, new) *UpdateResult`
+
+All the supporting validation code within kube-openapi will need to be refactored
+to support both creates and updates, and in the case of an update, to evaluate
+whether an error should be ignored.
+
+`ValidateUpdate`'s `UpdateResult` should return two separate lists: the
+hard validation errors, and any ignored errors.
+
+
+#### Nested Value Validations (anyOf, allOf)
+
+OpenAPI-schemas support nested validations:
+- not (disallowed)
+- oneOf (disallowed)
+- anyOf
+- allOf
+
+
+In the case of `anyOf` and `allOf`, we can support ratcheting these rules. The 
+schema validator will consider the match to be successful even if it has
+ignored errors due to ratcheting.
+
+### Structural-Schema-based validation changes
+
+The two structural-schema based validation methods will need to
+be refactored to support `ValidateUpdate` returning separate lists of errors.
+
+Alternatively, these could be rewritten into kube-openapi, and replaced with
+a validator that a schema is structural.
```

Comment:
We cannot do that. There is go-based validation deeply embedded in `objectmeta.Validate`.

In general, kube-openapi/pkg/validator is legacy code (taken from go-openapi). Moving anything there is very likely not what we want and we should have very good reasons to do that anyway.

Reply:
updated to remove this bit